src/dmd/dtemplate.d

```diff
@@ -4097,7 +4106,15 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                     result = MATCH.constant;
                     return;
                 }
+
+                uint oldatlock = t.aliasthislock;
+                t.aliasthislock |= AliasThisRec.RECtracing;
                 result = t.implicitConvTo(tp);
+                t.aliasthislock = oldatlock;
```

Comment:
This pattern of locking/unlocking that's repeated in many places should not be shouldered to the caller. Consider moving to a function.